crates/trie/db/src/state.rs

```diff
@@ -0,0 +1,238 @@
+use reth_db_api::transaction::DbTx;
+use reth_execution_errors::StateRootError;
+use reth_primitives::{BlockNumber, B256};
+use reth_trie::{
+    hashed_cursor::HashedPostStateCursorFactory, prefix_set::PrefixSetLoader, updates::TrieUpdates,
+    HashedPostState, StateRoot, StateRootProgress,
+};
+use std::ops::RangeInclusive;
+use tracing::debug;
+
+/// Extends [`StateRoot`] with operations specific for working with a database transaction.
+pub trait StateRootFromDbTx<'a, TX>: Sized {
+    /// Create a new [`StateRoot`] instance.
+    fn from_tx(tx: &'a TX) -> Self;
+
+    /// Given a block number range, identifies all the accounts and storage keys that
+    /// have changed.
+    ///
+    /// # Returns
+    ///
+    /// An instance of state root calculator with account and storage prefixes loaded.
+    fn incremental_root_calculator(
+        tx: &'a TX,
+        range: RangeInclusive<BlockNumber>,
+    ) -> Result<Self, StateRootError>;
+
+    /// Computes the state root of the trie with the changed account and storage prefixes and
+    /// existing trie nodes.
+    ///
+    /// # Returns
+    ///
+    /// The updated state root.
+    fn incremental_root(
+        tx: &'a TX,
+        range: RangeInclusive<BlockNumber>,
+    ) -> Result<B256, StateRootError>;
+
+    /// Computes the state root of the trie with the changed account and storage prefixes and
+    /// existing trie nodes collecting updates in the process.
+    ///
+    /// Ignores the threshold.
+    ///
+    /// # Returns
+    ///
+    /// The updated state root and the trie updates.
+    fn incremental_root_with_updates(
+        tx: &'a TX,
+        range: RangeInclusive<BlockNumber>,
+    ) -> Result<(B256, TrieUpdates), StateRootError>;
+
+    /// Computes the state root of the trie with the changed account and storage prefixes and
+    /// existing trie nodes collecting updates in the process.
+    ///
+    /// # Returns
+    ///
+    /// The intermediate progress of state root computation.
+    fn incremental_root_with_progress(
+        tx: &'a TX,
+        range: RangeInclusive<BlockNumber>,
+    ) -> Result<StateRootProgress, StateRootError>;
+}
+
+/// Extends [`HashedPostState`] with operations specific to working with a database transaction.
+pub trait HashedPostStateFromDbTx {
```

Comment:
this trait is a bit misleading as we do not construct hashed post state from database transaction, but rather compute state root with hashed post state as overlay. please move these methods to the trait above and rename them to `overlay_state_root` and `overlay_state_root_with_progress`

Reply:
Makes sense to me.

Since moving it to the trait above makes it implemented for `StateRoot` type, it would then be redundant naming it `StateRoot:overlay_state_root`, so I just named it `overlay` and `overlay_with_updates`.